### PR TITLE
finetune sdk

### PR DESCRIPTION
**Pull request by raphaelcristal**
This adds supports for finetunes to the sdk. It uses the newly created dory proxy to the backend.

How to use this
```python
import cohere
from cohere.custom_model_dataset import CsvDataset

client = cohere.Client('YOUR-API-KEY')

csv_data = CsvDataset(train_file="path_to_your_datset.csv", delimiter=",")
cm = client.create_custom_model(
    name="name-your-custom-model",
    dataset=csv_data,
    custom_model_type="GENERATIVE"
)
if cm.status == "READY":
  client.generate(prompt="here goes your prompt", model=cm.model_id)

# client.get_custom_model(cm.id)
# client.get_custom_model_by_name(cm.name)
# client.list_custom_models()
```

TODO:

- [x] change log
- [x] bump version

### CODE REVIEW

**Comment by sanderland** [comment]
Seems like you want to skip the consistency test on 'custom_model' in function name  (and document it)

**Comment by raphaelcristal** [comment]
> Seems like you want to skip the consistency test on 'custom_model' in function name (and document it)

Document in the readme? To mention that there is no async version of the finetune api?

**Comment by sanderland** [comment]
> > Seems like you want to skip the consistency test on 'custom_model' in function name (and document it)
> 
> Document in the readme? To mention that there is no async version of the finetune api?

Yep. Alternatively you could just have versions in AsyncClient that call super without having any async advantage.

**Comment by raphaelcristal** [comment]
> > > Seems like you want to skip the consistency test on 'custom_model' in function name (and document it)
> > 
> > 
> > Document in the readme? To mention that there is no async version of the finetune api?
> 
> Yep. Alternatively you could just have versions in AsyncClient that call super without having any async advantage.

Okay I added the Readme note. I feel like for me as user I would be quite surprised if the async call was not async at all.

**Comment by raphaelcristal** [comment]
Question:
Do I have to regenerate the lock file if I only updated the version of the package?

**Comment by raphaelcristal** [comment]
> > Question: Do I have to regenerate the lock file if I only updated the version of the package?
> 
> No, the lock file mainly specifies the exact version of dependencies for standalone use, which mainly affects CI tests and such.
> 
> > Okay I added the Readme note. I feel like for me as user I would be quite surprised if the async call was not async at all.
> 
> Keep in mind that AsyncClient inherits from Client, so now you probably have functions you can call, but they only crash. Maybe raise a NotImplementedError ?

Oh that's a valid point. I will add the exception.

File: cohere/client_async.py [hunk 1]
```diff
@@ -645,6 +659,134 @@ async def wait_for_bulk_embed_job(
...
+            json["settings"]["evalFiles"].append({"path": remote_path, **dataset.file_config()})
+
+        response = await self._request(f"{cohere.CUSTOM_MODEL_URL}/CreateFinetune", method="POST", json=json)
+        return CustomModel.from_dict(response["finetune"])
+
+    async def _upload_dataset(
+        self, content: Iterable[bytes], custom_model_name: str, file_name: str, type: INTERNAL_CUSTOM_MODEL_TYPE
+    ) -> str:
+        gcs = await self._create_signed_url(custom_model_name, file_name, type)
+        async with aiohttp.ClientSession() as session:
```
Comment (sanderland) [inline cohere/client_async.py:718]: would grab the existing session here

File: cohere/client_async.py [hunk 2]
```diff
@@ -645,6 +659,134 @@ async def wait_for_bulk_embed_job(
...
+            async with session.put(
+                url=gcs["url"], data=b"".join(content), headers={"content-type": "text/plain"}
```
[the remark this answers is not in the document]
Comment (sanderland) [reply cohere/client_async.py:720]: I believe you can pass a file handle

[the remark this answers is not in the document]
Comment (raphaelcristal) [reply cohere/client_async.py]: yeah the data might be already in memory though. anyways I won't optimise this further and just load it into memory.

**Review by sanderland — APPROVED** [review]

**Review by CoderHam — APPROVED** [review]
Loved the detailed function descriptors (with examples)

**Review by sanderland — APPROVED** [review]

**Pull request merged**